daemon/protocol/daemon.proto

```diff
@@ -27,31 +31,86 @@ message EchoReply {
   string pong = 1;
 }
 
-enum ProcessState {
-  UNKNOWN = 0;
-  PENDING = 1;
-  RUNNING = 2;
 
-  COMPLETED = 3;
-  FAILED = 4;
+// Create snapshot.
+//
+message CreateSnapshotRequest {
+  // Absolute path to a file or directory on the local filesystem.
+  string path = 1;
 }
 
-message Command {
-  repeated string argv = 1;
-  map<string, string> env = 2;
-  // Timeout in nanoseconds
-  int64 timeout = 3;
+message CreateSnapshotReply {
+  string error = 1;
+  string snapshot_id = 2;
 }
 
-message ProcessStatus {
-  string run_id = 1;
-  ProcessState state = 2;
-  string stdout_ref = 3;
-  string stderr_ref = 4;
-  int32 exit_code = 5;
-  string error = 6;
+
+// Checkout snapshot.
+//
+message CheckoutSnapshotRequest {
+  string snapshot_id = 1;
+
+  // Absolute path to a directory on the local filesystem (need not exist yet).
+  string dir = 2;
+}
+
+message CheckoutSnapshotReply {
+  string error = 1;
 }
 
-message StatusQuery {
+
+// Run
+//
+message RunRequest {
+  message Command {
+    repeated string argv = 1;
+    map<string, string> env = 2;
+    int64 timeout_ns = 3; //TODO: consistent special values for timeouts like PollRequest.timeout_ns?
+  }
+  message OutputPlan {
+    bool stdout = 1;
+    bool stderr = 2;
+
+    // After the run is done, generate a snapshot containing only the specified src paths.
+    // This copies the src files/dirs into the corresponding destination directories.
+    // The sources and destinations are all relative paths within the snapshot.
+    // An empty map will result in an empty snapshot (except for stdout/stderr).
+    // Behavior is undefined for duplicate entries within a destination.
+    // Sources may include the '*' wildcard.
+    map<string, string> src_paths_to_dest_dirs = 3;
+  }
+  string snapshot_id = 1;
+  Command cmd = 2;
+  OutputPlan plan = 3;
+}
+
+message RunReply {
   string run_id = 1;
-}
\ No newline at end of file
+  string error = 2;
```

Comment:
LGTM 
-one thought: do we want to return a status in RunReply?  or will the status always either be 'error' or 'starting'...
can the status from Run() be 'running'/'preparing' - it is starting immediately,  'pending' - it is queued behind other run requests?